### PR TITLE
Upgrade to django 1.11 and implemented tox

### PR DESCRIPTION
**Pull request by ahsan-ul-haq**
LEARNER-1529

### CODE REVIEW

**Comment by clintonb** [comment]
It would be ideal if the package were made Python 3-compatible. It's going to happen sooner, or later. Why not now?

File: search/urls.py [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
 COURSE_ID_PATTERN = getattr(settings, "COURSE_ID_PATTERN", r'(?P<course_id>[^/+]+(/|\+)[^/+]+(/|\+)[^/]+)')
 
 # urlpatterns is the standard name to use here
 # pylint: disable=invalid-name
-urlpatterns = patterns(
-    '',
+urlpatterns = [
     url(r'^$', views.do_search, name='do_search'),
     url(r'^{}$'.format(COURSE_ID_PATTERN), views.do_search, name='do_search'),
     url(r'^course_discovery/$', views.course_discovery, name='course_discovery'),
```
Comment (robrap) [inline search/urls.py]: Unrelated, but does anyone know what this course_discovery is used for?  Is it used by Open edX, or not at all? It was [added here](https://github.com/edx/edx-search/commit/35020548f27e87b663ed07bcd8de5f1d9da2abe1).  The only usages of edx-search I was aware of was course search (but I _think_ that is different), and team search.

**Comment by ahsan-ul-haq** [comment]
@clintonb yes that would be ideal but it would be out of scope for this ticket and after having a discussion with @macdiesel it appears that they have decided that it will be out of scope for this project.

**Comment by clintonb** [comment]
There's hardly any code in this repo. Test coverage is 100%. 2to3 does most of the conversion: https://docs.python.org/2/library/2to3.html. The scope of conversion seems the most minimal of nearly all of our libraries. If adding Python 3 support meant an extra week of time, I'd agree. However, this shouldn't take more than a couple hours.

**Comment by macdiesel** [comment]
@clintonb after a lot of deliberation the Platform team decided to not move forward on upgrading to python 3 at the same time as the 1.11 upgrade for the following reasons:
- On a small scale doing just this single repo is easy, now multiply this by the 20+ repos that are going to be touched and at a larger scale this becomes more difficult. 
- The philosophy of Doing one thing at a time well. Let's not inject additional complexity if we do not need to do so.

I do agree that this will be easy to upgrade to python 3, and I encourage teams to do so where they think it will be easy and they have the time. That said I would prefer they come in as a different version number on a separate PR. 

I do not wish to limit teams abilities to do what they think is right for their applications, however targeting python 3 support for any repository is not in scope for the django 1.11 project.

**Comment by ahsan-ul-haq** [comment]
@macdiesel just to clarify do we need upgrade to django 1.11 for this repo? Also we do need to add tox support. Please correct me if I am wrong.

**Comment by macdiesel** [comment]
@ahsan-ul-haq Yes, tox and django 1.11 support should be added to this repository.

**Comment by ahsan-ul-haq** [comment]
@macdiesel @clintonb @robrap please review.

File: search/tests/test_search_result_processor.py [hunk 2]
```diff
@@ -25,8 +25,8 @@ def test_strings_in_dictionary(self):
         })
         get_strings = SearchResultProcessor.strings_in_dictionary(test_dict)
         self.assertEqual(len(get_strings), 2)
-        self.assertEqual(get_strings[0], test_dict["a"])
```
Comment (clintonb) [inline search/tests/test_search_result_processor.py]: Why were the tests changed?

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    -rtest_requirements.txt
+
+commands =
+    coverage run ./manage.py test --settings=settings
+    coverage report
+
+[testenv:quality]
+commands =
+    pep8 --config=.pep8 search
+	pylint --rcfile=pylintrc search
```
Comment (clintonb) [inline tox.ini]: Unnecessary tab.

File: edxsearch/urls.py [hunk 4]
```diff
@@ -8,4 +9,7 @@
 
 # urlpatterns is the standard name to use here
 # pylint: disable=invalid-name
-urlpatterns = patterns('', url(r'^search/', include(search.urls)),)
+if django.VERSION < (1, 9):
+    urlpatterns = [url(r'^search/', include(search.urls))]
```
Comment (clintonb) [inline edxsearch/urls.py]: Django 1.9+ still supports the use of the `include()` function. It has simply been deprecated: https://docs.djangoproject.com/en/1.11/ref/urls/#include. There is no reason to fork here unless we decide, in the future, we must continue supporting Django < 1.11.

File: search/tests/test_search_result_processor.py (hunk 2 above)
Comment (ahsan-ul-haq) [inline search/tests/test_search_result_processor.py]: It seems like the list indexes does not remain in the same order so it was causing flakiness. I think all we need to check is that the strings are in the list and order of them does not matter. Correct me if I wrong.

Comment (clintonb) [inline search/tests/test_search_result_processor.py]: You're asking the wrong person. This package is simply something I had to touch to deal with the Elasticsearch upgrade. I have no idea what these tests are supposed to be doing.

**Comment by ahsan-ul-haq** [comment]
@clintonb made the suggested changes.

File: edxsearch/urls.py (hunk 4 above)
Comment (robrap) [inline edxsearch/urls.py]: Not a blocker, but I do wonder which is preferred...forking and dealing with everything at the same time and making the difference clear or leaving deprecation warnings with no clear indication in code why they weren't yet handled.

**Comment by robrap** [comment]
Do we want to test on a sandbox with edx-platform before merging or are we confident in relying on unit tests?

**Comment by ahsan-ul-haq** [comment]
@robrap I have created PR https://github.com/edx/edx-platform/pull/15541 and making sandbox to test things.

**Comment by ahsan-ul-haq** [comment]
@robrap I have created [sandbox](https://edx-search-testing.sandbox.edx.org) for testing can you help me with the testing.

**Comment by robrap** [comment]
@ahsan-ul-haq Looking at sandbox now...

**Comment by ahsan-ul-haq** [comment]
@robrap let me know if you have some issue.

**Pull request merged**